src/preferences/common/preference-service.ts

```diff
@@ -0,0 +1,127 @@
+import { Emitter, Event } from "../../application/common";
+import { inject } from "inversify";
+// import { inject, injectable } from "inversify";
+import { FileSystem, FileChangesEvent, FileStat } from "../../filesystem/common/filesystem";
+// import { FileSystemNode } from "../../filesystem/node/node-filesystem";
+
+
+
+import { FileSystemWatcher } from "../../filesystem/common/filesystem-watcher"
+
+export class Preference {
+    constructor(public key: string, public value: any) { }
+}
+export interface PreferenceChangedEvent {
+    readonly preferenceName: string;
+    readonly newValue?: any;
+    readonly oldValue?: any;
+}
+
+/**
+ * Minimal preference API with basic functionalities
+ */
+export interface PreferenceService {
+    // A minimal part of the interface.
+    // readonly preferenceNames: Promise<string[]>;
+
+    has(preferenceName: string): Promise<boolean>;
+    get<T>(preferenceName: string): Promise<T | undefined>;
+    // get<T>(preferenceName: string, defaultValue?: T): Promise<T>;
+    // set<T>(preferenceName: string, value: T): void;
+    // remove<T>(preferenceName: string): T | undefined;
+
+}
+
+export interface PreferenceServiceClient {
+    onDidChangePreference(event: PreferenceChangedEvent): void
+}
+
+export class JsonPreferenceService implements PreferenceService {
+
+    // A minimal part of the interface.
+    readonly onDidChangePreference: Event<PreferenceChangedEvent>;
+    prefs: Promise<any>; // Preferences cache
+    wsRoot: Promise<string>;
+    private prefsChangedListeners: Emitter<PreferenceChangedEvent> = new Emitter();
+
+    constructor(
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
+        readonly prefFileRelativeUri: string) {
+        this.wsRoot = fileSystem.getWorkspaceRoot().then((fileStat: FileStat) => {
+            return Promise.resolve(fileStat.uri);
+        })
+
+        watcher.onFileChanges(event => {
+            this.arePreferencesAffected(event).then((arePrefsAffected) => {
+                if (arePrefsAffected) {
+                    this.reconcilePreferences();
+                }
+            })
+        });
+
+        this.reconcilePreferences();
+    }
+
+    /**
+     * Checks to see if the preference file was modified
+     */
+    arePreferencesAffected(event: FileChangesEvent): Promise<boolean> {
+        return new Promise<boolean>((resolve) => {
+            this.wsRoot.then((root) => {
+                for (let change of event.changes) {
+                    if (change.uri === this.wsRoot + "/" + this.prefFileRelativeUri) {
+                        resolve(true);
+                    }
+                }
+                resolve(false);
+            })
+        })
+    }
+
+    /**
+     * Read preferences
+     */
+    reconcilePreferences(): void {
+        this.wsRoot.then((root) => {
+            const uri = root + "/" + this.prefFileRelativeUri;
```

Comment:
Since I am working with the paths on Windows in general... can we have a URI instead. Then you can append segments. In worst case, please use 
```
import * as path from "path";

// the later
path.resolve(root, whetever).
```

But I would go with URI append. The current one clearly will not work on Windows.